src/oatpp/network/Server.cpp

```diff
@@ -45,28 +46,48 @@ Server::Server(const std::shared_ptr<ServerConnectionProvider> &connectionProvid
 void Server::mainLoop(Server *instance) {
 
   instance->setStatus(STATUS_STARTING, STATUS_RUNNING);
-
   std::shared_ptr<const std::unordered_map<oatpp::String, oatpp::String>> params;
 
-  while (instance->getStatus() == STATUS_RUNNING) {
-
-    auto connection = instance->m_connectionProvider->get();
-
-    if (connection) {
-      if (instance->getStatus() == STATUS_RUNNING) {
-        instance->m_connectionHandler->handleConnection(connection, params /* null params */);
+  // Code duplication to extract unnecessary checks for the conditional to maintain performance when the conditional isn't used.
+  if (instance->m_condition) {
+    while (instance->getStatus() == STATUS_RUNNING) {
+      if (instance->m_condition()) {
+        auto connection = instance->m_connectionProvider->get();
+
+        if (connection) {
+          if (instance->getStatus() == STATUS_RUNNING) {
+            if (instance->m_condition()) {
+              instance->m_connectionHandler->handleConnection(connection, params /* null params */);
+            } else {
+              instance->setStatus(STATUS_STOPPING);
+            }
+          } else {
+            OATPP_LOGD("[oatpp::network::server::mainLoop()]", "Error. Server already stopped - closing connection...");
+          }
+        }
       } else {
-        OATPP_LOGD("[oatpp::network::server::mainLoop()]", "Error. Server already stopped - closing connection...");
+        instance->setStatus(STATUS_STOPPING);
+      }
+    }
+  } else {
+    while (instance->getStatus() == STATUS_RUNNING) {
+      auto connection = instance->m_connectionProvider->get();
+
+      if (connection) {
+        if (instance->getStatus() == STATUS_RUNNING) {
+          instance->m_connectionHandler->handleConnection(connection, params /* null params */);
+        } else {
+          OATPP_LOGD("[oatpp::network::server::mainLoop()]", "Error. Server already stopped - closing connection...");
+        }
       }
     }
-
   }
 
   instance->setStatus(STATUS_DONE);
 
 }
 
-void Server::run(bool startAsNewThread) {
+void Server::run(bool startAsNewThread, std::function<bool()> conditional) {
```

Comment:
I would prefer to get rid of a `startAsNewThread` - it's too confusing to me.
Let `Server::run()` always execute on a current thread.



```cpp
void Server::run(std::function<bool()> stopCondition = nullptr);
```

Reply:
Its a feature we introduced in 1.1 iirc since we received a ton of issues because

> `server.run()` never returns

We have not received any complaints since then. My vote would be for keeping it. Maybe we should do some kind of poll in gitter/github?

Reply:
Yes, I know. I treated that mostly as a workaround. 
To my mind, it doesn't give any benefits. All it does is creating a new thread that you could've created on your own and adds a ton of complexity to the "Server" class.

Also, I understand why it was previously a bit confusing for some people (blocking forever), however with this current PR, it seems to me that it may fix that confusion:

- If no condition passed - it's blocked forever (until stop is called)
- If there is a condition - it's blocked until the condition is met.

The `startAsNewThread` parameter here is redundant and blows up the complexity and the interface.

Reply:
Ok, now I get it and I can understand your point. We can phase that out. I have an nice solution for this in my mind. I try to push it until tomorrow.